packages/components/src/popover/index.js

```diff
@@ -159,30 +196,29 @@ const Popover = (
 	}, [ anchorRef, anchorRect, getAnchorRect ] );
 
 	/**
-	 * Offsets the position of the popover when the anchor is inside an iframe.
+	 * Adjusts the position of the popover when the anchor is inside an iframe
+	 * so that the popover aligns correctly with its anchor.
 	 */
-	const frameOffset = useMemo( () => {
-		const { defaultView } = ownerDocument;
-		const { frameElement } = defaultView;
+	const iframeOffsetMiddleware = useMemo( () => {
+		const frameElement = ownerDocument?.defaultView?.frameElement;
 
 		if ( ! frameElement || ownerDocument === document ) {
 			return undefined;
 		}
 
-		const iframeRect = frameElement.getBoundingClientRect();
-		return {
-			name: 'iframeOffset',
-			fn( { x, y } ) {
-				return {
-					x: x + iframeRect.left,
-					y: y + iframeRect.top,
-				};
-			},
-		};
+		const { top, left } = frameElement.getBoundingClientRect();
+
+		return offsetMiddleware( ( { placement } ) => {
+			const computedMainAxis = getMainAxisFromPlacement( placement );
+			return {
+				mainAxis: computedMainAxis === 'x' ? -top : -left,
```

Comment:
I think there are a couple of fixes to be made here. If the main axis is `x` (horizontal), the offset should be `left` (also horizontal).

I think the problem is that `getMainAxisFromPlacement` is returning the wrong value. From the [docs](https://floating-ui.com/docs/offset#mainaxis), the main axis is the one on which the distance between the reference and floating elements would be measure, so for a `top` or `bottom` placement you'd measure along the `y` axis.

I also noticed in #42417 that for `top` and `left` placements, the offset needs to be subtracted on the main axis, and for `right` and `bottom` placements added.

When that's all addressed, the code will be very similar to #42417, so perhaps it's worth reviewing/merging that PR? I tested it pretty thoroughly on a range of placements.

Reply:
> When that's all addressed, the code will be very similar to #42417, so perhaps it's worth reviewing/merging that PR? I tested it pretty thoroughly on a range of placements.

That is definitely true, although this PR also adds a more refined behavior to `limitShift`, and several other minor improvements.

Probably the best course of action is to review / merge #42417 and then rebase this PR and refine the `limitShift` approach ?

Reply:
Sounds good to me, thanks 👍